packages/web-extension/test/remoteApi/remoteApi.integration.test.ts

```diff
@@ -474,4 +512,73 @@ describe('remoteApi integration', () => {
       });
     });
   });
+
+  describe('factory', () => {
+    let destructor: StubDestructor;
+
+    beforeEach(() => {
+      destructor = new StubDestructor();
+      sut = setUp({ destructor });
+      sut.emitInitial();
+    });
+
+    afterEach(() => {
+      sut.cleanup();
+      createSubjects.cache.clear!();
+    });
+
+    it('accessing property returns the same function object', () => {
+      expect(sut.consumer.factory).toBe(sut.consumer.factory);
+      expect(sut.consumer.nested.factory).toBe(sut.consumer.nested.factory);
+    });
+
+    const testFactory = (
+      factoryName: string,
+      getFactory: (consumer: typeof sut.consumer) => typeof consumer.factory
+    ) => {
+      describe(factoryName, () => {
+        it('accessing property returns the same function object', () => {
+          expect(getFactory(sut.consumer)).toBe(getFactory(sut.consumer));
+        });
+
+        it('calling method on returned remote api calls remote method and resolves the result', async () => {
+          const api = getFactory(sut.consumer)();
+          expect(await api.addOne(2n)).toBe(3n);
+        });
+
+        it('garbage collected consumed api shuts down the messengers', (done) => {
+          const areFactoryMessengersShutdown = () => {
+            // This is likely to break on lodash update
+            // eslint-disable-next-line @typescript-eslint/no-explicit-any
+            const [_, subjects] = Object.entries((createSubjects.cache as any).__data__.string.__data__).find(([key]) =>
+              key.includes('-factory')
+            )!;
+            const { consumerMessage$, hostMessage$ } = subjects as ReturnType<typeof createSubjects>;
+            return !consumerMessage$.observed && !hostMessage$.observed;
+          };
+          // Create and consume a new remote api through another API's factory method
+          // eslint-disable-next-line promise/always-return
+          void Promise.resolve(getFactory(sut.consumer)()).then(() => {
+            expect(destructor.callbacks).toHaveLength(1);
+            setTimeout(() => {
+              expect(areFactoryMessengersShutdown()).toBe(false);
+              // Emulate GC freeing the object returned by the factory method (consume side)
+              destructor.callbacks[0]();
+              setTimeout(() => {
+                expect(areFactoryMessengersShutdown()).toBe(true);
+                // eslint-disable-next-line promise/no-callback-in-promise
+                done();
+              }, 1);
+            }, 1);
+          });
+        });
+
+        // Don't know how to test this
```

Comment:
Can it be covered with an integration test?

Reply:
Could potentially be implemented as e2e test as we'd want it to run in the intended environment (web extension). I verified that it works manually as writing a test would take a lot of effort.